zcash_client_sqlite/src/wallet.rs

```diff
@@ -873,58 +918,113 @@ fn subtree_scan_progress(
             .flatten()
             .map(Ok)
             .or_else(|| {
-                conn.query_row(
-                    &format!(
-                        "SELECT MAX({table_prefix}_commitment_tree_size - {output_count_col})
-                        FROM blocks
-                        WHERE height <= :start_height",
-                    ),
-                    named_params![":start_height": u32::from(birthday_height)],
+                stmt_start_tree_size
+                    .query_row(
+                        named_params![":start_height": u32::from(birthday_height)],
+                        |row| row.get::<_, Option<u64>>(0),
+                    )
+                    .optional()
+                    .map(|opt| opt.flatten())
+                    .transpose()
+            })
+            .transpose()?;
+
+        // Get the note commitment tree size as of the end of the recover-until height.
+        let recover_until_size = recover_until_height
+            .map(|end_height| {
+                stmt_start_tree_size
+                    .query_row(
+                        named_params![":start_height": u32::from(end_height + 1)],
+                        |row| row.get::<_, Option<u64>>(0),
+                    )
+                    .optional()
+                    .map(|opt| opt.flatten())
+            })
+            .transpose()?;
+
+        // Count the total outputs scanned so far on the birthday side of the
+        // recover-until height.
+        let recovered_count = recover_until_height
```

Comment:
`recover_until_height` is optional (it can be set to `None` when constructing an `AccountBirthday`), so `recovered_count` is therefore necessarily also optional.

(Figuring out all the different bits of data in our database that may or may not be there is why this took me a solid day to implement.)